caffe2/core/storage.h

```diff
@@ -16,52 +16,55 @@
 #include "caffe2/core/logging.h"
 #include "caffe2/core/typeid.h"
 
+#include <ATen/core/Allocator.h>
 #include <ATen/core/Device.h>
 #include <ATen/core/DeviceType.h>
 #include <ATen/core/intrusive_ptr.h>
 
 namespace caffe2 {
 
-using DataType = TypeMeta;
-using DataPtr = std::shared_ptr<void>;
-using at::DeviceType;
-
 class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
  public:
   StorageImpl() = delete;
   StorageImpl(const StorageImpl&) = delete;
   StorageImpl& operator=(const StorageImpl&) = delete;
 
-  explicit StorageImpl(DeviceType device_type) : device_type_(device_type) {}
-  StorageImpl(DeviceType device_type, TypeMeta data_type)
-      : data_type_(data_type), device_type_(device_type) {}
-  template <typename Deleter = MemoryDeleter>
+  // Rule of Five
+  StorageImpl(StorageImpl&&) = default;
+  ~StorageImpl() = default;
+  StorageImpl& operator=(StorageImpl&&) = default;
+
   StorageImpl(
-      DeviceType device_type,
       TypeMeta data_type,
-      void* src,
-      size_t capacity,
-      Deleter d = nullptr)
-      : data_type_(data_type), device_type_(device_type) {
-    CAFFE_ENFORCE_WITH_CALLER(
-        data_type_.id() != TypeIdentifier::uninitialized(),
-        "To create storage with a raw external pointer you need to pass in an "
-        "initialized data_type(TypeMeta).");
-    // Check if the deleter is a MemoryDeleter and is a simple nullptr.
-    if (std::is_same<MemoryDeleter, Deleter>::value &&
-        reinterpret_cast<MemoryDeleter*>(static_cast<void*>(&d))[0] ==
-            nullptr) {
-      // Use aliasing constructor trick to avoid calling the destructor.
-      data_ptr_ = std::shared_ptr<void>(std::shared_ptr<void>(), src);
-    } else {
-      data_ptr_.reset(src, d);
+      int64_t numel,
+      at::DataPtr data_ptr,
+      at::Allocator* allocator,
+      bool resizable)
+      : data_type_(data_type),
+        data_ptr_(std::move(data_ptr)),
+        numel_(numel),
+        resizable_(resizable),
+        allocator_(allocator) {
+    if (numel > 0) {
+      CAFFE_ENFORCE(
+          data_type.id() != TypeIdentifier::uninitialized(),
+          "Constructing a storage with meta of unknown type and non-zero numel");
```

Comment:
nit: The old error message seemed better; it told you how to fix the problem, as opposed to this one, which just tells you there's a problem